[develop] Make export-logs able to download the logs using the default/custom PCluster bucket as staging bucket

hehe7318 opened the pull request:
### Description of changes
- The CLI commands `export-cluster-logs` and `export-image-logs` can now by default export the logs to the default ParallelCluster bucket or to the CustomS3Bucket if specified in the config.
- If not specified `--bucket` option, cannot export logs to `parallelcluster/` as it is a protected folder.
- When by default using the pcluster default bucket(`parallelcluster-****************-v1-do-not-delete`), it will automatically update the bucket policy to the latest one to adapt the changes.
- Pcluster default bucket bootstrapped_file format, content and verify method changed.
- Unit tests and integration tests for the new changes.

### Tests
* Unit test fixed and passed. Added new unit tests for new functions. Also covered some old functions.
* Adapted changes for integ tests. Tests worked as expected.
* Manually test done and worked as expected

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hehe7318 committed:
Make export-logs able to download the logs using the default/custom pcluster bucket as staging bucket.

hehe7318 committed:
Correct an error, using self.name instead of get_bucket_name

hehe7318 committed:
Reformat, remove unused import

hehe7318 committed:
Corrected some errors.

gmarciani commented on CHANGELOG.md:
```diff
@@ -3,14 +3,12 @@ CHANGELOG
 ------
 
+**CHANGES**
+- export-cluster-logs/export-image-logs now is able to download the logs using the default/custom PCluster bucket as staging bucket
```
I'd rephrase to "The CLI commands `export-cluster-logs` and `export-image-logs` can now export the logs to the default ParallelCluster bucket and the `CustomS3Bucket` if no other bucket is specified."

hehe7318 replied:
Done!

gmarciani commented on cli/src/pcluster/cli/commands/cluster_logs.py:
```diff
@@ -38,8 +38,10 @@ def register_command_args(self, parser: ArgumentParser) -> None:  # noqa: D102
-            help="S3 bucket to export cluster logs data to. It must be in the same region of the cluster",
+            help=(
+                "S3 bucket to export cluster logs data to. It must be in the same region of the cluster. "
+                "If not specified, the default/custom ParallelCluster S3 bucket will be used."
```
I'd rephrase to "If not specified, the ParallelCluster default bucket or the CustomS3Bucket (if specified in the config) will be used."

hehe7318 replied:
Done!

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -43,8 +43,10 @@ def register_command_args(self, parser: ArgumentParser) -> None:  # noqa: D102
-            help="S3 bucket to export image builder logs data to. It must be in the same region of the image",
+            help=(
+                "S3 bucket to export image builder logs data to. It must be in the same region of the image. "
+                "If not specified, the default ParallelCluster S3 bucket will be used."
```
Why can't we use the CustomS3Bucket bucket as well (if specified), the same way we do for clusters?

hehe7318 replied:
We can, but when doing manually tests I made some mistakes and it mislead me. Fixed

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -58,20 +60,35 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
                 self._validate_output_file_path(args.output_file)
-            return self._export_image_logs(args, args.output_file)
+
+            is_cluster_bucket = False if args.bucket else True
```
The name `is_cluster_bucket` is misleading. What about `is_pcluster_bucket`?
This variable is used to determine if a prefix validation must be done or not; it just be done only if the pcluster bucket is used.

hehe7318 replied:
Good catch! Thank you. Fixed

gmarciani commented on cli/src/pcluster/cli/commands/cluster_logs.py:
```diff
@@ -65,18 +67,33 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
                 self._validate_output_file_path(args.output_file)
-            return self._export_cluster_logs(args, args.output_file)
+
+            is_cluster_bucket = False if args.bucket else True
```
The name `is_cluster_bucket` is misleading. What about `is_pcluster_bucket`?
This variable is used to determine if a prefix validation must be done or not; it just be done only if the pcluster bucket is used.

hehe7318 replied:
Fixed

gmarciani commented on cli/src/pcluster/cli/commands/cluster_logs.py:
```diff
@@ -65,18 +67,33 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
-    def _export_cluster_logs(args: Namespace, output_file: str = None):
+    def _validate_bucket_prefix(bucket_prefix: str) -> None:
+        if bucket_prefix:
+            if bucket_prefix.startswith("parallelcluster/") or bucket_prefix=="parallelcluster":
```
**CodeStyle** Magic strings should be avoided as much as magic numbers.
I suggest to define a constant `PCLUSTER_BUCKET_PROTECTED_PREFIXES` and use it for the checks

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/cli/commands/cluster_logs.py:
```diff
@@ -65,18 +67,33 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
+        if bucket_prefix:
+            if bucket_prefix.startswith("parallelcluster/") or bucket_prefix=="parallelcluster":
+                raise ValueError(
+                    "Cannot export logs to the protected folder 'parallelcluster/'. Please use another folder."
```
The fact that some folders are protected when the pcluster bucket is used must be reflected in the option helper and the public doc.

Also the error message must be more specific and dynamic:  "Cannot export logs to `{bucket_prefix}` as it is a protected folder in {default bucket}. Please use another folder."

hehe7318 replied:
Done. Also add msg in `--bucket-prefix` help panel.

gmarciani commented on cli/src/pcluster/cli/commands/cluster_logs.py:
```diff
@@ -65,18 +67,33 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
+                )
+
+    @staticmethod
+    def _export_cluster_logs(args: Namespace, output_file: str = None, is_cluster_bucket: bool = False):
```
Why do we need to add `is_cluster_bucket`?
The line below may be 

`bucket=args.bucket if arg.bucket else cluster.bucket`

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -58,20 +60,35 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
         url = imagebuilder.export_logs(
-            bucket=args.bucket,
+            # imagebuilder.bucket will handle the bucket init including policy update
+            bucket=imagebuilder.bucket.name if is_cluster_bucket else args.bucket,
```
Why do we need to add `is_cluster_bucket`?
The line below may be 

`bucket=args.bucket if arg.bucket else image_builder.bucket`

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -58,20 +60,35 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
+        if bucket_prefix:
+            if bucket_prefix.startswith("parallelcluster/") or bucket_prefix=="parallelcluster":
+                raise ValueError(
+                    "Cannot export logs to the protected folder 'parallelcluster/'. Please use another folder."
```
Same comment made for the cluster export

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -154,6 +154,95 @@ def configure_s3_bucket(self):
+        account_id = self.account_id
+        bucket_arn = f"arn:{partition}:s3:::{self.name}"
+        bucket_arn_with_wildcard = f"{bucket_arn}/*"
+        logs_service_principal = f"logs.{region}.{get_url_domain_suffix(partition)}"
```
This is a good candidate to be a common function because there are other places in our code where we create service principal strings. The function may be: `def get_service_principal(service_name: str, region: str, regional: bool = False)`.

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -154,6 +154,95 @@ def configure_s3_bucket(self):
+                    "StringEquals": {"s3:x-amz-acl": "bucket-owner-full-control", "aws:SourceAccount": account_id},
+                    "ArnLike": {
+                        "aws:SourceArn": [
+                            f"arn:{partition}:logs:{region}:{account_id}:log-group:/aws/parallelcluster/*",
```
this is a good candidate for a common function. See `format_arn()` https://github.com/aws/aws-parallelcluster/blob/62fdb5eefbb866139180c7795ce539913506b218/cli/src/pcluster/templates/awsbatch_builder.py#L98-L99

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -154,6 +154,95 @@ def configure_s3_bucket(self):
+        required_statements = self._get_bucket_policy_for_cloudwatch_logs()
+
+        # Check which statements are missing
+        existing_sids = {stmt.get("Sid") for stmt in existing_policy.get("Statement", []) if "Sid" in stmt}
```
With this logic, SIDs cannot be changed in the bucket policy and also it's a a weak check: we are interested in statement content, not in statement ids. 
anyway, this check may not be necessary, I suggest to overwrite the bucket policy without checking the SIDs because we are using the pcluster bucket that we are fully in control of.

hehe7318 replied:
Done.

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -354,6 +443,13 @@ def _check_default_bucket(cls, service_name: str, artifact_directory: str, stack
+        else:
+            # Bucket exists and is bootstrapped, ensure bucket policy is up-to-date
+            try:
+                bucket.ensure_bucket_policy_for_logs()
```
Improvement, that can be done in a follow up PR.

We should configure the bucket policy within `_configure_bucket()` where we are already setting the DenyHTTP statement https://github.com/aws/aws-parallelcluster/blob/8ac2c636f7941bc8983ee75f6abf202028b2d031/cli/src/pcluster/models/s3_bucket.py#L144-L156

At that point we must change the logic to check if the bucket is already bootstrapped:
1. change the bootstrap file body [here](https://github.com/aws/aws-parallelcluster/blob/8ac2c636f7941bc8983ee75f6abf202028b2d031/cli/src/pcluster/models/s3_bucket.py#L191) to reflect what are the bootstrapped features required, e.g. basic (what we already have up to Pc 3.11.1) and export-logs (introduced in PC 3.12.0). 
2. if the bootstrap file does not exist or it does not contain all the required features, then overwrite the bucket policy

hehe7318 replied:
With the changes removing SIDs check, we will need this logic in this PR. Need some more time to implement

hehe7318 replied:
Done.

hehe7318 committed:
Address comments and fix unit tests

- Extract common functions get_service_principal and format_arn to reduce code duplication.
- Simplify S3 bucket policy updates by directly generating and setting the complete policy.
- Modify bootstrap file handling to include a list of bootstrapped features, ensuring backward compatibility.
- Update logic to reconfigure the bucket only when required features are missing.
- Optimization of some variable names
- Help panel message update.
- Error message update.
- CHANGELOG.md update.
- Fix 2 unit tests to adapt the new changes.

hehe7318 committed:
Minor changes on format.

hehe7318 committed:
Minor changes on format.

hehe7318 committed:
Minor changes on format.

gmarciani commented on CHANGELOG.md:
```diff
@@ -3,14 +3,12 @@ CHANGELOG
 ------
 
+**CHANGES**
+- The CLI commands `export-cluster-logs` and `export-image-logs` can now export the logs to the default ParallelCluster bucket and the `CustomS3Bucket` if no other bucket is specified.
```
nit: Re-reading I think the line is confusing. 
We could replicate here the same sentence used in the helper: "to the default ParallelCluster bucket or to the `CustomS3Bucket` if specified in the config".

hehe7318 replied:
Done! Thanks!

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -43,26 +43,48 @@ def register_command_args(self, parser: ArgumentParser) -> None:  # noqa: D102
+            ):
+                raise ValueError(
+                    f"Cannot export logs to {bucket_prefix} as it is a protected folder "
+                    f"in {PCLUSTER_BUCKET_PROTECTED_PREFIX}. Please use another folder."
```
as it is *within* a protected folder

hehe7318 replied:
Done, also done for cluster_logs.py

gmarciani replied:
I mean `Cannot export logs to {bucket_prefix} as it is within the protected folder {PCLUSTER_BUCKET_PROTECTED_PREFIX}. Please use another folder.`

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -43,26 +43,48 @@ def register_command_args(self, parser: ArgumentParser) -> None:  # noqa: D102
+            is_pcluster_bucket = not args.bucket
+
+            if is_pcluster_bucket:
+                self._validate_bucket_prefix(args.bucket_prefix)
```
the prefix validation should be done both on the pcluster default bucket and on the CustomS3Bucket. The variable name should better reflect it, or at least add a comment here explaining when the validation is performed. Otherwise, it seems that the validation is done only for the default pcluster bucket.

hehe7318 replied:
Done, added a comment to explain the scenario.

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -184,19 +270,41 @@ def delete_s3_artifacts(self):
+                bootstrapped_features = bootstrapped_info.get("bootstrapped_features", [])
+            except json.JSONDecodeError:
+                LOGGER.info("Bootstrap file is in old format. Assuming basic features are bootstrapped.")
+                bootstrapped_features = ["basic"]
```
Let's define bootstrap features as constants

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -184,19 +270,41 @@ def delete_s3_artifacts(self):
+            return True
+        except AWSClientError as e:
+            if e.error_code == "404":
+                LOGGER.info("Bootstrap file doesn't exist. S3 Bucket %s is not configured.", self.name)
```
This line should be a warning

hehe7318 replied:
Done. BTW, it's a normal scenario when the cluster is created as first time.

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -184,19 +270,41 @@ def delete_s3_artifacts(self):
+                bootstrapped_info = json.loads(file_content)
+                bootstrapped_features = bootstrapped_info.get("bootstrapped_features", [])
+            except json.JSONDecodeError:
+                LOGGER.info("Bootstrap file is in old format. Assuming basic features are bootstrapped.")
```
Let's add a comment here explaining what's the old format like

hehe7318 replied:
Done. Added # The bootstrap file old format is a simple String "bucket is configured successfully."

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -184,19 +270,41 @@ def delete_s3_artifacts(self):
+                LOGGER.info("Bootstrap file doesn't exist. S3 Bucket %s is not configured.", self.name)
+                return False
+            else:
+                LOGGER.error("Unable to check S3 bucket %s is configured properly.", self.name)
```
typo: "...to check if..."

hehe7318 replied:
Done, added `if`.

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -370,17 +479,13 @@ def _create_bucket(cls, bucket: S3Bucket, e: AWSClientError):
+    def _configure_bucket(cls, bucket: S3Bucket):
+        try:
+            bucket.configure_s3_bucket()
+            bucket.upload_bootstrapped_file(bootstrapped_features=bucket.REQUIRED_BOOTSTRAP_FEATURES)
```
Why should the function `upload_bootstrapped_file` accept argument `bootstrapped_features`?
Is there any scenario where the list of features should be different from `REQUIRED_BOOTSTRAP_FEATURES `?
If the answer is no, then let's remove the argument.

hehe7318 replied:
Good catch. Thank you.

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/constants.py:
```diff
@@ -299,3 +299,6 @@ class Operation(Enum):
 IAM_INSTANCE_PROFILE_REGEX = "^arn:.*:instance-profile/"
 IAM_POLICY_REGEX = "^arn:.*:policy/"
+
+PCLUSTER_BUCKET_PROTECTED_PREFIX = "parallelcluster/"
```
Let's interpolated PCLUSTER_BUCKET_PROTECTED_FOLDER rather than repeating `parallelcluster`

hehe7318 replied:
Done

hehe7318 committed:
Merge branch 'develop' into wip/export-logs-enhancement

hehe7318 committed:
Address comments

hehe7318 committed:
Fix a bug that when bootstrapped doesn't exist it cannot correctly catch 404 error code by execute head_object first. Fix unit tests. Add new unit test test_check_bucket_is_bootstrapped.

gmarciani commented on cli/tests/pcluster/cli/test_export_cluster_logs/TestExportClusterLogsCommand/test_helper/pcluster-help.txt:
```diff
@@ -18,11 +18,16 @@ options:
                         Keypath under which exported logs data will be stored
                         in s3 bucket. Defaults to <cluster_name>-logs-<current
-                        time in the format of yyyyMMddHHmm>
+                        time in the format of yyyyMMddHHmm>, If not specify
```
Typo: "if not specify" => "if not specified"

hehe7318 replied:
Done

gmarciani commented on cli/tests/pcluster/cli/test_export_cluster_logs/TestExportClusterLogsCommand/test_helper/pcluster-help.txt:
```diff
@@ -18,11 +18,16 @@ options:
-                        time in the format of yyyyMMddHHmm>
+                        time in the format of yyyyMMddHHmm>, If not specify
+                        `--bucket` option, cannot export logs to
+                        parallelcluster/ as it is a protected folder.
```
Let's use the constant here instead of `parallelcluster/`

hehe7318 replied:
This is a txt file for the unit test.

gmarciani commented on cli/tests/pcluster/cli/test_export_image_logs/TestExportImageLogsCommand/test_helper/pcluster-help.txt:
```diff
@@ -33,8 +34,13 @@ options:
                         Keypath under which exported logs data will be stored
                         in s3 bucket. Defaults to <image_id>-logs-<current
-                        time in the format of yyyyMMddHHmm>
+                        time in the format of yyyyMMddHHmm>. If not specify
```
Same as above for the typo and the use of constant

hehe7318 replied:
Done for the typo. But this is a txt file for the unit test.

gmarciani commented on cli/tests/pcluster/models/dummy_s3_bucket.py:
```diff
@@ -87,6 +87,7 @@ def mock_bucket_object_utils(
     delete_s3_artifacts_side_effect=None,
     upload_bootstrapped_file_side_effect=None,
     check_bucket_is_bootstrapped_side_effect=None,
+    check_bucket_is_bootstrapped_return_value=None,
```
why the check should return None, rather than True/False?

hehe7318 replied:
Fixed. Changed to True. Thanks for catching!

gmarciani commented on cli/tests/pcluster/cli/test_commands.py:
```diff
@@ -333,3 +343,101 @@ def test_setup_bucket_with_resources_upload_failure(
+        "expected_exception",
+    ),
+    [
+        # Test case 1: The bootstrap file exists, content is in the new format, and contains all required features
```
Rather than describing the test with a comment, use the test id.
See example here: https://github.com/aws/aws-parallelcluster/blob/c84baa2955744e418eccc42c93c339131ef36a06/cli/tests/pcluster/aws/test_ec2.py#L168

hehe7318 replied:
Done, thanks for knowledge sharing.

gmarciani commented on cli/src/pcluster/utils.py:
```diff
@@ -579,3 +579,19 @@ def retrieve_supported_regions():
     return retrieve_supported_regions.cache
+
+
+def get_service_principal(service_name: str, partition: str, region: str = None, regional: bool = False) -> str:
```
This new function should be covered by a unit test

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/utils.py:
```diff
@@ -579,3 +579,19 @@ def retrieve_supported_regions():
+        return f"{service_name}.{domain_suffix}"
+
+
+def format_arn(partition: str, service: str, region: str, account: str, resource: str) -> str:
```
This new function should be covered by a unit test

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -148,12 +148,96 @@ def configure_s3_bucket(self):
+
+        return bucket_policy
+
+    def _get_bucket_policy_for_cloudwatch_logs(self):
```
We should cover the generation of the policy with a unit test

hehe7318 replied:
Done, but seems like it's unnecessary, due to it doesn't have logic that worth to test. The generated policy statements depend on the return value of `get_service_principal`, `format_arn`, etc.

gmarciani replied:
A unit test is necessary even if the function returns a static string because it protects from regressions.

Example: what if I change the bucket policy version? That string is static, but still I want my unit test to capture the mistake.

Makes sense?

hehe7318 replied:
Makes sense

gmarciani commented on cli/src/pcluster/models/s3_bucket.py:
```diff
@@ -148,12 +148,96 @@ def configure_s3_bucket(self):
+        bucket_policy = self._generate_bucket_policy()
+        AWSApi.instance().s3.put_bucket_policy(bucket_name=self.name, policy=json.dumps(bucket_policy))
+
+    def _generate_bucket_policy(self):
```
We should cover the generation of this bucket policy with a unit test. I know this policy exist before your PR, but it's worth it to take the chance to cover it.

hehe7318 replied:
Done

gmarciani commented on cli/src/pcluster/cli/commands/image_logs.py:
```diff
@@ -43,26 +43,51 @@ def register_command_args(self, parser: ArgumentParser) -> None:  # noqa: D102
             "--bucket-prefix",
             help="Keypath under which exported logs data will be stored in s3 bucket. Defaults to "
-            "<image_id>-logs-<current time in the format of yyyyMMddHHmm>",
+            "<image_id>-logs-<current time in the format of yyyyMMddHHmm>. If not specify `--bucket` option, "
```
Same as above: typo

hehe7318 replied:
Done

gmarciani reviewed (changes requested):
We should implement a logic that prevents the deletion of objects from the protected folders

hehe7318 committed:
Address comments. Add unit tests for new codes.

hehe7318 commented:
> We should implement a logic that prevents the deletion of objects from the protected folders

I can't see why. We don't have a objects deletion logic in our codebase. Also we didn't add logic that put objects in protected folder. 

I suppose even if we should implement this logic it should not be included in this PR.

hehe7318 committed:
Add unit test for not valid bucket-prefix

hehe7318 committed:
Adapt changes in integ tests

gmarciani commented on cli/src/pcluster/cli/commands/cluster_logs.py:
```diff
@@ -65,18 +70,40 @@ def execute(self, args: Namespace, extra_args: List[str]) -> None:  # noqa: D102
             return None
 
+    @staticmethod
+    def _validate_bucket_prefix(bucket_prefix: str) -> None:
```
This function is equal to the one used for the image builder part. Can we remove code duplication?

hehe7318 replied:
Done

gmarciani commented on cli/tests/pcluster/cli/test_commands.py:
```diff
@@ -333,3 +343,117 @@ def test_setup_bucket_with_resources_upload_failure(
+        ),
+    ],
+)
+def test_check_bucket_is_bootstrapped(
```
This unit test should be moved to `cli/tests/pcluster/models/test_s3_bucket.py` because the original function 
`check_bucket_is_bootstrapped()` is in file `cli/src/models/s3_bucket`.

hehe7318 replied:
Done, thanks for catching this.

gmarciani commented on cli/tests/pcluster/test_utils.py:
```diff
@@ -607,3 +607,114 @@ async def async_method(self, param):
+@pytest.mark.parametrize(
+    "partition, service, region, account, resource, expected_result",
+    [
+        pytest.param(
```
*Testing Best Practice*
The function under test is joining input values following a given schema, but it does not make any assumption on the input values (and this is fine).

In this case, the test parameters should be generic:

Example: 

```
 pytest.param(
            "aws",
            "s3",
            "",
            "123456789012",
            "bucket_name",
            "arn:aws:s3::123456789012:bucket_name",
            id="Empty region",
        ),
``` 

should be:

```
 pytest.param(
            "ANY_PARTITION",
            "ANY_SERVICE",
            "",
            "ANY_ACCOUNT_ID",
            "ANY_BUCKET_NAME",
            "arn:ANY_PARTITION:ANY_SERVICE::ANY_ACCOUNT_ID:ANY_BUCKET_NAME",
            id="Empty region",
        ),
```

hehe7318 replied:
Done.

hehe7318 committed:
Merge branch 'develop' into wip/export-logs-enhancement

hehe7318 committed:
Addressed comments

gmarciani commented on cli/tests/pcluster/test_utils.py:
```diff
@@ -607,3 +607,114 @@ async def async_method(self, param):
+def test_get_service_principal(mocker, service_name, partition, region, regional, expected_result, expected_exception):
+    """Test get_service_principal with various inputs."""
+    # Mock get_url_domain_suffix
+    domain_suffix = "amazonaws.com.cn" if partition == "aws-cn" else "amazonaws.com"
```
*Test Best Practice*
This if..else logic is not necessary.
Here you're unit testing the function `get_service_principal`, which does not make any assumption on the value returned by `get_url_domain_suffix`. So the mock should return a static value `AWS_DOMAIN_SUFFIX` and the test should assert that the function `get_url_domain_suffix`is called once

hehe7318 replied:
You are right, I should not mock `get_url_domain_suffix`. I re-checked the code and I decided to remove the mock logic and let the function itself handle the `domain_suffix`. Does it make sense to you?

gmarciani replied:
It's ok, let's not block the Pr for this.
However, the best practice would be to let the get_url_domain_suffix return a static value `A_PARTITION` and the expected result is adapted accordigly.

hehe7318 replied:
Done

hehe7318 committed:
Addressed comments

gmarciani commented on cli/tests/pcluster/models/test_s3_bucket.py:
```diff
@@ -301,3 +304,241 @@ def test_upload_file(mocker, content, file_name, file_type, s3_file_format, expe
+
+    policy_statements = bucket._get_bucket_policy_for_cloudwatch_logs()
+
+    assert_that(policy_statements).is_length(3)
```
You could have simplified this long set of assertions with a single assert that the policy is equal to the expected json, without asserting field by field.

hehe7318 replied:
Done

gmarciani commented on cli/tests/pcluster/models/test_s3_bucket.py:
```diff
@@ -301,3 +304,241 @@ def test_upload_file(mocker, content, file_name, file_type, s3_file_format, expe
+    assert_that(deny_put_object_statement["Principal"]).is_equal_to({"Service": logs_service_principal})
+
+
+def test_generate_bucket_policy(mocker):
```
Same as above: reduce the assertion to an assert equal to a static json instead of asserting field by field.

hehe7318 replied:
Done!

hehe7318 commented:
> > We should implement a logic that prevents the deletion of objects from the protected folders
> 
> I can't see why. We don't have a objects deletion logic in our codebase. Also we didn't add logic that put objects in protected folder.
> 
> I suppose even if we should implement this logic it should not be included in this PR.

Had a sync with Giacomo, we have validator and DenyPutS3Object(only for logs service) policy for the default pcluster bucket. So although we have objects deletion logic in our codebase(--keep-s3-objects false), it doesn't matter.

hehe7318 committed:
Improve unit tests

hehe7318 committed:
Improve unit tests

gmarciani committed:
Merge branch 'develop' into wip/export-logs-enhancement
